segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ThrottlerCalculator.java

```diff
@@ -65,6 +65,16 @@
      */
     @VisibleForTesting
     static final double DURABLE_DATALOG_THROTTLE_THRESHOLD_FRACTION = 0.1;
+    /**
+     * Maximum size (in number of operations) of the OperationLog, above which maximum throttling will be applied.
+     */
+    @VisibleForTesting
+    static final int OPERATION_LOG_MAX_SIZE = 1_000_000;
```

Comment:
It would be good to experiment whether this throttler comes into play at the right time (e.g., not too early for very small appends or too late for large ones). Actually, as the objective of this is not to be left without cache space when a recovery occurs, it would be better to throttle based on the number of bytes of the operation log in relation to the cache size configured (I don't know if this is possible).

Reply:
Well, there's always a balance between one thing and another. The objective of this throttler is to prevent the Tier 1 log from getting too big, which would cause unnecessarily long delays during recovery. 1 Million operations to recover is reasonably large buffer. The recovery speed is a function of both the read throughput from BK and the number of operations processed. Each operation has to be deserialized, written to the cache, its effects applied to metadata, etc.

If we do some quick math:
- If all we had were 100 byte appends, 1M x 100 byte is about 100MB. That should have been no problem for the StorageWriter to dump to LTS. If it had trouble doing that prior to the recovery, then we likely had another issue which had to be resolved.
    - For a 10GB cache with 4 containers on a pod, that is about 2.5GB cache per container, so this means a backlog of 25 million operations to recover. We would prefer not to get into this situation to begin with.
- For 10KB appends (more likely, due to client-side batching), 1M x 10KB is about 10GB. At this point we should get parity with whatever cache size we set (assuming only one container on the box - if multiple containers, then parity would happen much earlier).